client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1524 @@
+		if self.drop_reserved_peer(&peer_id)? || self.drop_regular_peer(&peer_id) {
+			// The peer found and disconnected.
+			self.report_disconnect(peer_id, reason);
+			Ok(())
+		} else {
+			// The peer was not found in neither regular or reserved lists.
+			Err(peer_id)
+		}
+	}
+
+	/// Try dropping the peer as a reserved peer. Return `Ok(true)` if the peer was found and
+	/// disconnected, `Ok(false)` if it wasn't found, `Err(PeerId)`, if the peer found, but not in
+	/// connected state.
+	fn drop_reserved_peer(&mut self, peer_id: &PeerId) -> Result<bool, PeerId> {
+		let Some(state) = self.reserved_nodes.get_mut(peer_id) else {
+			return Ok(false)
+		};
+
+		if let PeerState::Connected(_) = state {
+			*state = PeerState::NotConnected;
+			Ok(true)
+		} else {
+			Err(peer_id.clone())
+		}
+	}
+
+	/// Try dropping the peer as a regular peer. Return `true` if the peer was found and
+	/// disconnected, `false` if it wasn't found.
+	fn drop_regular_peer(&mut self, peer_id: &PeerId) -> bool {
+		let Some(direction) = self.nodes.remove(peer_id) else {
+			return false
+		};
+
+		match direction {
+			Direction::Inbound => self.num_in -= 1,
+			Direction::Outbound => self.num_out -= 1,
+		}
+
+		true
+	}
+
+	/// Initiate outgoing connections trying to connect all reserved nodes and fill in all outgoing
+	/// slots.
+	fn alloc_slots(&mut self) {
+		// Try connecting to reserved nodes first.
+		self.reserved_nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				(matches!(state, PeerState::NotConnected) && !self.peer_store.is_banned(*peer_id))
+					.then(|| {
+						*state = PeerState::Connected(Direction::Outbound);
+						peer_id
+					})
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.into_iter()
+			.for_each(|peer_id| {
+				self.start_connection(peer_id);
+			});
+
+		// Nothing more to do if we're in reserved-only mode or don't have slots available.
+		if self.reserved_only || self.num_out >= self.max_out {
+			return
+		}
+
+		// Fill available slots with non-reserved nodes
+		let available_slots = (self.max_out - self.num_out).saturated_into();
+		// Ignore reserved nodes (connected above) and already connected nodes.
+		let ignored = self
+			.reserved_nodes
+			.keys()
+			.collect::<HashSet<&PeerId>>()
+			.union(&self.nodes.keys().collect::<HashSet<&PeerId>>())
+			.cloned()
+			.collect();
+
+		self.peer_store
+			.outgoing_candidates(available_slots, ignored)
+			.iter()
+			.filter_map(|peer_id| {
+				if self.reserved_nodes.contains_key(peer_id) || self.nodes.contains_key(peer_id) {
+					debug_assert!(false, "`PeerStore` returned a node we asked to ignore.");
+					error!(
+						target: "peerset",
+						"`PeerStore` returned a node we asked to ignore: {}.",
+						peer_id
+					);
+					None
+				} else {
+					self.num_out += 1;
+					self.nodes.insert(*peer_id, Direction::Outbound);
+					Some(peer_id)
+				}
+			})
+			.take(available_slots)
```

Comment:
why is this needed?

Reply:
This is a precaution in case `peer_store` misbehaves and returns more records than requested. May be I should also report an error in this case.

Reply:
If I'm reading the code correctly, this precaution doesn't apply to the code that converts the peer into a connected peer. Maybe it would be better to split this code into two: first fetch the candidates and the connect to them one by one. Having this now split into two makes it more error prone I think.

Reply:
Yeah, I also spotted this issue. Should be fixed by https://github.com/paritytech/substrate/pull/13611/commits/463653674527ae9d7206150f09dd8d540ab670a5